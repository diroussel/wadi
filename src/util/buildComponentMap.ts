import {glob} from 'glob';
import type {ComponentMap} from '../types/gully-types';
import {getComponents} from './getComponents';

export async function buildComponentMap(
	addLocalZipPath: boolean,
	projRootDir: string,
): Promise<ComponentMap> {
	const componentList = await getComponents(projRootDir);

	const componentMap: ComponentMap = Object.fromEntries(
		await Promise.all(
			componentList.map(async ({componentName, packagePath}) => {
				if (addLocalZipPath) {
					const localPath = await glob(
						`${packagePath}/target/dist/DistPrefix.${componentName}-*.zip`,
					);
					if (localPath.length !== 1) {
						throw new Error(`Local zip not found, packagePath: ${packagePath}`);
					}

					return [componentName, {localZipFile: localPath[0]}];
				}

				return [componentName, {}];
			}),
		),
	);
	return componentMap;
}
