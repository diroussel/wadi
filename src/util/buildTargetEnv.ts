import type { TargetEnvParams } from '../commands/common-args/target-env-args';
import type { TargetEnvFile } from '../types/gully-types';

export function buildTargetEnv(params: TargetEnvParams): TargetEnvFile {
  const {
    componentMap,
    componentVersion,
    versionRoot,
    branch,
    s3AssetsBucket,
    s3AssetsPathRoot,
  } = params;

  // Find the build_type, in the future this should some from the component artifact-metadata.json
  function attributes(component: string) {
    let type = '';
    if (component.endsWith('static')) {
      type = 'static_asset';
    } else if (
      [
        'backchannellogoutlambda',
        'mockidplambda',
        'igflambda',
        'landingpagelambda',
        'bsamocklambda',
        'viewerapplambda',
        'candidatesdataapi',
        'notifycallbacklambda',
      ].includes(component)
    ) {
      type = 'apigateway';
    } else if (
      component.endsWith('lambda') ||
      component === 'dynamodb2sqs' ||
      component.includes('canary')
    ) {
      // TODO: rename dynamodb2sqs
      type = 'lambda';
    } else if (['gully'].includes(component)) {
      type = 'pipelineutils';
    } else if (
      component.endsWith('test') ||
      component.endsWith('e2e') ||
      component.endsWith('postdeploy')
    ) {
      type = 'postdeploy';
    } else if (component === 'terraform' || component === 'pipelinelocking') {
      type = 'deploy_sh';
    }
    return { build_type: type };
  }

  const components = Object.fromEntries(
    Object.entries(componentMap)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([component, componentData]) => [
        component,
        {
          ...attributes(component),
          ...componentData,
          version: componentVersion,
          zip_prefix: `NHSD.PopulationHealth.${component}-`,
          zip_branch: branch,
        },
      ])
  );

  return {
    version_root: versionRoot,
    s3_assets_bucket: s3AssetsBucket,
    s3_assets_path_root: s3AssetsPathRoot,
    components,
  };
}
