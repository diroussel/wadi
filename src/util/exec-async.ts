import process from 'node:process';
import {exec, type ExecOptions} from 'node:child_process';

export type AsyncExecOptions = {
	log?: boolean;
};

// Async wrapper around child_process.exec()
export async function execAsync(
	command: string,
	options: AsyncExecOptions & ExecOptions = {},
): Promise<{stdout: string; stderr: string}> {
	const log = options?.log ?? process.env.DEBUG_EXEC === 'true';
	if (log) {
		console.error(command);
	}

	return new Promise((done, failed) => {
		exec(command, {...options}, (error, stdout, stderr) => {
			if (log) {
				console.error(`Output from ${command} was:\n${stdout}`);
			}

			if (error) {
				failed(error);
				return;
			}

			done({stdout, stderr});
		});
	});
}

/**
 * Read a string from the stdout of another process
 */
export async function execStr(
	command: string,
	options: AsyncExecOptions & ExecOptions = {},
): Promise<string> {
	const proc = await execAsync(command, options);
	return proc.stdout.trim();
}
