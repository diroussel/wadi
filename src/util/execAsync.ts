/* eslint-disable no-param-reassign,no-console */
import { exec, type ExecOptions } from 'child_process';

export type AsyncExecOptions = {
  log?: boolean;
};

// async wrapper around child_process.exec()
export async function execAsync(
  command: string,
  options: AsyncExecOptions & ExecOptions = {}
): Promise<{ stdout: string; stderr: string }> {
  const log = options?.log || process.env.DEBUG_EXEC === 'true';
  if (log) console.error(command);
  return new Promise((done, failed) => {
    exec(command, { ...options }, (err, stdout, stderr) => {
      if (log) {
        console.error(`Output from ${command} was:\n${stdout}`);
      }
      if (err) {
        failed(err);
        return;
      }
      done({ stdout, stderr });
    });
  });
}

/**
 * Read a string from the stdout of another process
 */
export async function execStr(
  command: string,
  options: AsyncExecOptions & ExecOptions = {}
): Promise<string> {
  return (await execAsync(command, options)).stdout.trim();
}
