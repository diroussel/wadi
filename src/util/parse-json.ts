import {readFile} from 'node:fs/promises';
import type {TargetEnvFile} from '../types/wadi-types';

/**
 * Read some JSON and cast to the supplied type. Not very safe, but safe enough
 */
export function parseJson<T>(buffer: Uint8Array): T {
	return JSON.parse(new TextDecoder().decode(buffer)) as T;
}

export async function readJsonFile<T>(path: string): Promise<T> {
	return parseJson<T>(await readFile(path));
}

export async function readTargetEnvJson(
	targetEnvJsonPath: string,
): Promise<TargetEnvFile> {
	const targetEnvJson = parseJson<TargetEnvFile>(
		await readFile(targetEnvJsonPath),
	);

	if (!targetEnvJson.components) {
		throw new Error(
			`"components" element not found in json, see '${targetEnvJsonPath}'`,
		);
	}

	return targetEnvJson;
}
