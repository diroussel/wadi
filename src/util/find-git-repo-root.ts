import {execStr} from './exec-async';

export async function findGitRepoRoot() {
	return await execStr('git rev-parse --show-toplevel');
}
