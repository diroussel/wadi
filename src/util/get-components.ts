import path from 'node:path';
import {glob} from 'glob';
import pMap from 'p-map';
import type {PkgJson, ComponentList, ComponentLocation} from '../types/wadi-types';
import {readJsonFile} from './parse-json';

async function listComponentsInPackage(packageJsonPath: string): Promise<ComponentList> {
	const fileData = await readJsonFile<PkgJson>(packageJsonPath);
	const packagePath = path.dirname(packageJsonPath);

	const names = fileData.componentNames ?? (fileData.componentName ? [fileData.componentName] : []);
	return names.map(componentName => ({componentName, packagePath}));
}

export async function getComponents(
	projRootDir: string,
): Promise<ComponentList> {
	const rootPkgJson = await readJsonFile<PkgJson>(
		`${projRootDir}/package.json`,
	);
	if (!rootPkgJson?.workspaces) {
		throw new Error('workspaces not found');
	}

	const workspaces = rootPkgJson.workspaces.join(',');
	const paths = await glob(`${projRootDir}/{${workspaces}}/package.json`, {
		ignore: '**/node_modules/**',
	});

	const list = await pMap(paths, listComponentsInPackage, {concurrency: 3});
	return list.flat();
}
