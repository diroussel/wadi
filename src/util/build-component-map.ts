import pMap from 'p-map';
import {glob} from 'glob';
import type {ComponentList, ComponentLocation, ComponentMap} from '../types/wadi-types';
import {getComponents} from './get-components';

export async function buildComponentMap(
	addLocalZipPath: boolean,
	projRootDir: string,
): Promise<ComponentMap> {
	const findLocalZip = async ({componentName, packagePath}: ComponentLocation): Promise<[string, {localZipFile?: string}]> => {
		if (!addLocalZipPath) {
			return [componentName, {}];
		}

		const localPath = await glob(
			`${packagePath}/target/dist/DistPrefix.${componentName}-*.zip`,
		);
		if (localPath.length === 0) {
			throw new Error(`Local zip not found, packagePath: ${packagePath}`);
		}

		if (localPath.length !== 1) {
			throw new Error(`Too many zips found, packagePath: ${packagePath}`);
		}

		return [componentName, {localZipFile: localPath[0]}];
	};

	const componentList = await getComponents(projRootDir);
	return Object.fromEntries(await pMap(componentList, findLocalZip, {concurrency: 4}));
}
