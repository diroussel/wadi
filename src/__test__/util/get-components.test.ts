import {glob} from 'glob';
import {getComponents} from '../../util/get-components';
import {readJsonFile} from '../../util/parse-json';
import type {PkgJson} from '../../types/wadi-types';

jest.mock('../../util/parse-json');
jest.mock('glob');
const readJsonFileMock = jest.mocked(readJsonFile);
const globMock = jest.mocked(glob);

describe('getComponents', () => {
	it('an error is thrown if no workspaces can be read from root package.json', async () => {
		await expect(
			async () => getComponents('rootDir'),
		).rejects.toThrowError('workspaces not found');
	});

	it('glob should be called with workspaces, and projRootDir arg', async () => {
		readJsonFileMock.mockResolvedValueOnce({
			workspaces: ['workspaceDir1', 'workspaceDir2'],
		});

		globMock.mockResolvedValueOnce([]);

		await getComponents('testRootDirectory');

		expect(glob).toBeCalledWith(
			'testRootDirectory/{workspaceDir1,workspaceDir2}/package.json',
			{ignore: '**/node_modules/**'},
		);
	});

	it('readJsonFile will be called with each path returned by glob', async () => {
		readJsonFileMock.mockImplementation(async (path): Promise<PkgJson> => {
			switch (path) {
				case 'rootDir/package.json': {return {
					version: '1.0.0',
					workspaces: ['workspaceDir1'],
				};}

				case 'first/path/package.json': {return {
					version: '1.0.0',
					componentNames: ['frontendlambda', 'frontendstatic'],
				};}

				case 'and/second/path/package.json': {return {
					version: '1.0.0',
					componentName: 'app1lambda',
				};}

				default: {
					throw new Error(`Wasn't expecting path ${path}`);
				}
			}
		});
		globMock.mockResolvedValueOnce([
			'first/path/package.json',
			'and/second/path/package.json',
		]);

		const components = await getComponents('rootDir');
		expect(components).toMatchSnapshot();

		expect(readJsonFile).toBeCalledWith('first/path/package.json');
		expect(readJsonFile).toBeCalledWith('and/second/path/package.json');
	});

	it('if packageJson has no componentNames, it is not included in returned array', async () => {
		readJsonFileMock.mockResolvedValueOnce({
			workspaces: ['workspaceDir1'],
		});
		globMock.mockResolvedValueOnce([
			'first/path/package.json',
			'and/second/path/package.json',
		]);
		readJsonFileMock.mockResolvedValueOnce({
			componentNames: ['component'],
		});
		readJsonFileMock.mockResolvedValueOnce({});

		expect(await getComponents('rootDir')).toEqual([
			{componentName: 'component', packagePath: 'first/path'},
		]);
	});

	it('packages with multiple names result in multiple components with the same path', async () => {
		readJsonFileMock.mockResolvedValueOnce({
			workspaces: ['workspaceDir1'],
		});
		globMock.mockResolvedValueOnce(['first/path/package.json']);
		readJsonFileMock.mockResolvedValueOnce({
			componentNames: ['firstComponent', 'andSecondComponent'],
		});

		expect(await getComponents('rootDir')).toEqual([
			{componentName: 'firstComponent', packagePath: 'first/path'},
			{componentName: 'andSecondComponent', packagePath: 'first/path'},
		]);
	});
});
