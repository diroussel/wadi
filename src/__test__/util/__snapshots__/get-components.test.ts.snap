// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponents readJsonFile will be called with each path returned by glob 1`] = `
[
  {
    "componentName": "frontendlambda",
    "packagePath": "first/path",
  },
  {
    "componentName": "frontendstatic",
    "packagePath": "first/path",
  },
  {
    "componentName": "app1lambda",
    "packagePath": "and/second/path",
  },
]
`;
